src/basic_solver.cpp

```diff
@@ -0,0 +1,166 @@
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> ATA_small(A_shape[1],A_shape[1]);
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> K_cross(N, N);
+    Eigen::Vector<T, Eigen::Dynamic> ATb_small(A_shape[1]);
+
+    // loop over each Star
+    for (ssize_t c=0; c < AList.size(); ++c) {
+        // build b
+        py::buffer_info b_buffer_info = bList[c].request();
+        T __restrict *b_data = static_cast<T *>(b_buffer_info.ptr);
+
+        // build A
+        py::buffer_info A_buffer_info = AList[c].request();
+        T __restrict *A_data = static_cast<T *>(A_buffer_info.ptr);
+
+        //build K
+        py::buffer_info K_buffer_info = KList[c].request();
+        T __restrict *K_data = static_cast<T *>(K_buffer_info.ptr);
+
+        // set some variables for sizes
+        ssize_t A_shape_one = A_buffer_info.shape[1];
+        ssize_t A_shape_zero = A_buffer_info.shape[0];
+
+        // calculate the outer product of the K parameters so that it can be re-used in the
+        // following loops
+        Eigen::Map<const Eigen::Vector<T, Eigen::Dynamic>> K_vec(K_data, N);
+        K_cross = K_vec *K_vec.transpose();
+
+        Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>> A_map(A_data, A_shape_zero, A_shape_one);
+        // reset the ATA array to all zeros
+        ATA_small.setZero();
+        // since the array will be self-adjoint, use Eigen such that only half the number
+        // of comutations need done.
+        ATA_small.template selfadjointView<Eigen::Upper>().rankUpdate(A_map.transpose());
+
+        Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 1>> b_map(b_data, b_buffer_info.shape[0]);
+        ATb_small = A_map.transpose() * b_map;
+    
+        // Calculate ATb contribution for this list entry
+        for (ssize_t o = 0; o < A_shape_one; o ++) {
+            ssize_t base_pos = o*N;
+            for (ssize_t l = 0; l < N; l++) {
+                ATb_data[base_pos+l] += K_data[l] * ATb_small(o);
+            }
+        }
+
+        // Calculate ATA contribution for this list entry
+        for (ssize_t j = 0; j < A_shape_one; j++) {
+            ssize_t base_pos_j = j*N;
+            for (ssize_t i = 0; i < j+1; i++) {
+                ssize_t base_pos = i*N;
+                ATA_eig.template block<N, N>(base_pos, base_pos_j) += ATA_small(i,j)*K_cross;
+            }
+        }
+    }
+
+    // solve Ax=b. If the matrix can be decomposed using llt do that (faster) if it cant be
+    // then use the slower ldlt method
+    Eigen::Map<const Eigen::Vector<T, Eigen::Dynamic>> ATb_map(ATb_data, A_shape[1]*N);
+    auto view = ATA_eig. template selfadjointView<Eigen::Upper>();
+    auto lltDecomp = view.llt();
+    Eigen::Matrix<T, Eigen::Dynamic, 1> result;
+    if (lltDecomp.info() == Eigen::ComputationInfo::Success) {
+        result = lltDecomp.solve(ATb_map);
+    } else {
+        auto decomp = view.ldlt();
+        result = decomp.solve(ATb_map);
```

Comment:
If the matrix is singular, this will also fail.  In the python code, we try to do something graceful in that case, switching to SVD.  I suspect we want something similar here.  Or at least a try/catch somewhere to inform the user what happened rather than let through a probably baroque Eigen error message.